xarray/coding/times.py

```diff
@@ -752,11 +815,84 @@ def encode_cf_datetime(
         # we already covered for this in pandas-based flow
         num = cast_to_int_if_safe(num)
 
-    return (num, units, calendar)
+    if dtype is not None:
+        num = _cast_to_dtype_if_safe(num, dtype)
+
+    return num, units, calendar
+
+
+def _encode_cf_datetime_within_map_blocks(
+    dates: np.ndarray,
+    units: str,
+    calendar: str,
+    dtype: np.dtype,
+) -> np.ndarray:
+    num, *_ = _eagerly_encode_cf_datetime(
+        dates, units, calendar, dtype, allow_units_modification=False
+    )
+    return num
+
+
+def _lazily_encode_cf_datetime(
+    dates: DaskArray,
+    units: str | None = None,
+    calendar: str | None = None,
+    dtype: np.dtype | None = None,
+) -> tuple[DaskArray, str, str]:
+    import dask.array
+
+    if calendar is None:
+        # This will only trigger minor compute if dates is an object dtype array.
+        calendar = infer_calendar_name(dates)
+
+    if units is None and dtype is None:
+        if dates.dtype == "O":
+            units = "microseconds since 1970-01-01"
+            dtype = np.dtype("int64")
+        else:
+            units = "nanoseconds since 1970-01-01"
+            dtype = np.dtype("int64")
+
+    if units is None or dtype is None:
+        raise ValueError(
+            f"When encoding chunked arrays of datetime values, both the units "
+            f"and dtype must be prescribed or both must be unprescribed. "
+            f"Prescribing only one or the other is not currently supported. "
+            f"Got a units encoding of {units} and a dtype encoding of {dtype}."
+        )
+
+    num = dask.array.map_blocks(
```

Comment:
Ah that's neat!  Is there anything special I need to do to test this or is it sufficient to demonstrate that it at least works with dask (as my tests already do)?

One other minor detail is how to handle typing with this.  I gave it a shot (hopefully I'm in the right ballpark), but I still encounter this error:
```
xarray/coding/times.py: note: In function "_lazily_encode_cf_datetime":
xarray/coding/times.py:860: error: "T_ChunkedArray" has no attribute "dtype"  [attr-defined]
```
I'm not a typing expert so any guidance would be appreciated!

Reply:
> I need to do to test this or is it sufficient to demonstrate that it at least works with dask (as my tests already do)

I believe the dask tests are fine since the cubed tests live elsewhere. 

> error: "T_ChunkedArray" has no attribute "dtype"  [attr-defined

cc @TomNicholas

Reply:
> I believe the dask tests are fine since the cubed tests live elsewhere.

Yep.

> Is there anything special I need to do

The correct `ChunkManager` for the type of the chunked array needs to be detected, see [this pattern]().

> error: "T_ChunkedArray" has no attribute "dtype" [attr-defined

This can't be typed properly yet (because we don't yet have a full protocol to describe the `T_DuckArray` in `xarray.core.types.py`), but try changing the definition of `T_ChunkedArray` in `parallelcompat.py` to this

```python
T_ChunkedArray = TypeVar("T_ChunkedArray", bound=Any)
```

Reply:
Gotcha, thanks for the typing guidance @TomNicholas.  Adding `bound=Any` to the `T_ChunkedArray` definition indeed solves the `dtype` attribute issue.  It seems I am not able to use `T_DuckArray` as an input type for a function, however — I'm getting errors like this:
```
error: Cannot use a covariant type variable as a parameter  [misc]
```
Do you have any thoughts on how to handle that?  I'm guessing the `covariant=True` parameter was added for a reason in its definition?

https://github.com/pydata/xarray/blob/41d33f52f709a765fb0dbfb5b9b4f5ea55173053/xarray/core/types.py#L176

Does https://github.com/pydata/xarray/pull/8575#discussion_r1439921180 make the output side of this more difficult to handle as well (at least if we want to be as accurate as possible)?

Reply:
@TomNicholas do you have any thoughts on this?  From my perspective the only thing holding this PR up at this point is typing.

Reply:
IMO a `type: ignore` is fine.

Reply:
Hey, sorry for forgetting about this @spencerkclark ! 

> I'm guessing the covariant=True parameter was added for a reason in its definition?

Yes - that fixes some typing errors in the `ChunkManager` methods IIRC. Unfortunately I don't yet understand how to make that work and also fix your error 🙁 

> the only thing holding this PR up at this point is typing.

I agree with Deepak that typing should not hold this up. Even if it needs some `ignore`s then the typing you've added is still very useful to indicate intent!